substrate/frame/parameters/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+//! trait, that allows to pass data along to the origin check. It gets passed in the key. The
+//! implementor can then match on the key and the origin to decide whether the origin is
+//! permissioned to set the value.
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+
+use frame_support::traits::{
+	dynamic_params::{AggregratedKeyValue, IntoKey, Key, RuntimeParameterStore, TryIntoKey},
+	EnsureOriginWithArg,
+};
+
+mod benchmarking;
+#[cfg(test)]
+mod tests;
+mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// The key type of a parameter.
+type KeyOf<T> = <<T as Config>::RuntimeParameters as AggregratedKeyValue>::AggregratedKey;
+
+/// The value type of a parameter.
+type ValueOf<T> = <<T as Config>::RuntimeParameters as AggregratedKeyValue>::AggregratedValue;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config(with_default)]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		#[pallet::no_default_bounds]
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The overarching KV type of the parameters.
+		///
+		/// Usually created by [`frame_support::dynamic_params`] or equivalent.
+		#[pallet::no_default_bounds]
+		type RuntimeParameters: AggregratedKeyValue;
+
+		/// The origin which may update a parameter.
+		///
+		/// The key of the parameter is passed in as second argument to allow for fine grained
+		/// control.
+		#[pallet::no_default_bounds]
+		type AdminOrigin: EnsureOriginWithArg<Self::RuntimeOrigin, KeyOf<Self>>;
+
+		/// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+
+		/// Provides a default KV since the type is otherwise in-constructable.
+		#[pallet::no_default]
+		#[cfg(feature = "runtime-benchmarks")]
+		type BenchmarkingDefault: Get<Self::RuntimeParameters>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A Parameter was set.
+		Updated {
+			/// The Key-Value pair that was set.
+			key_value: T::RuntimeParameters,
+		},
+	}
+
+	/// Stored parameters.
+	#[pallet::storage]
+	pub type Parameters<T: Config> =
+		StorageMap<_, Blake2_128Concat, KeyOf<T>, ValueOf<T>, OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set the value of a parameter.
+		///
+		/// The dispatch origin of this call must be `AdminOrigin` for the given `key`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_parameter())]
+		pub fn set_parameter(
+			origin: OriginFor<T>,
+			key_value: T::RuntimeParameters,
+		) -> DispatchResult {
+			let (key, value) = key_value.clone().into_parts();
+
+			T::AdminOrigin::ensure_origin(origin, &key)?;
+
+			Parameters::<T>::mutate(key, |v| *v = value);
+
+			Self::deposit_event(Event::Updated { key_value });
```

Comment:
This event will be deposited even if the value is the same as the current in storage. I'd prevent this from happening in case the `value ==  v`.

Reply:
See comment above, i now include the old value.